tests/src/rsa_test_helpers.c

```diff
@@ -0,0 +1,41 @@
+/* We're in test code so it's OK to access private funtions and structures' parameters. */
+#define MBEDTLS_DECLARE_PRIVATE_IDENTIFIERS
+#define MBEDTLS_ALLOW_PRIVATE_ACCESS
+
+#include "test/rsa_test_helpers.h"
+#include "mbedtls/private/rsa.h"
+#include "mbedtls/private/bignum.h"
+#include "test/bignum_helpers.h"
+#include "rsa_alt_helpers.h"
+
+int mbedtls_rsa_test_fill_context(mbedtls_rsa_context *rsa,
+                                  const unsigned char *P_data, size_t P_len,
+                                  const unsigned char *Q_data, size_t Q_len,
+                                  const unsigned char *N_data, size_t N_len,
+                                  const unsigned char *E_data, size_t E_len)
+{
+    mbedtls_mpi N; mbedtls_mpi P; mbedtls_mpi Q;
+    mbedtls_mpi E; mbedtls_mpi D;
+    int ret;
+
+    mbedtls_mpi_init(&N); mbedtls_mpi_init(&P); mbedtls_mpi_init(&Q);
+    mbedtls_mpi_init(&E); mbedtls_mpi_init(&D);
+
+    MBEDTLS_MPI_CHK(mbedtls_mpi_read_binary(&P, P_data, P_len));
+    MBEDTLS_MPI_CHK(mbedtls_mpi_read_binary(&Q, Q_data, Q_len));
+    MBEDTLS_MPI_CHK(mbedtls_mpi_read_binary(&N, N_data, N_len));
+    MBEDTLS_MPI_CHK(mbedtls_mpi_read_binary(&E, E_data, E_len));
+
+    MBEDTLS_MPI_CHK(mbedtls_rsa_deduce_private_exponent(&P, &Q, &E, &D));
+
+    MBEDTLS_MPI_CHK(mbedtls_rsa_import(rsa, &N, &P, &Q, &D, &E));
+
+#if !defined(MBEDTLS_RSA_NO_CRT)
+    MBEDTLS_MPI_CHK(mbedtls_rsa_deduce_crt(&P, &Q, &D, &rsa->DP, &rsa->DQ, &rsa->QP));
```

Comment:
Hmm, design question: is this helper function supposed to work with 3.6? If so, this line breaks in builds with `RSA_ALT`: in such builds, the context is not guaranteed to have fields names `DP` etc.

If the function is not supposed to work with 3.6, then this should be documented (the declaration should be guarded appropriately). But then we can get rid of the intermediate variables: directly `mpi_read_binary()` into `rsa->P` etc. then skip the call to `rsa_import()` (just set `rsa->len` directly).

If the function is supposed to work with 3.6, then we probably want separate implementations for 3.6 and 1.0+. The 3.6 implementation should rely strictly on `import()` + `complete()` and not access internals directly.